### PR TITLE
Multithreading support

### PR DESCRIPTION
**Pull request by matthewclegg**
This PR adds support for multithreading.  To use multithreading, run experiment_driver.py with the parameter --cores=0.

### CODE REVIEW

File: src/driver/experiment_driver.py [hunk 1]
```diff
@@ -102,20 +106,22 @@ def __init__(
...
         random_seed: int,
+        cores: int,
     ):
         self._data_design_dir = data_design_dir
         self._experimental_design = experimental_design
         self._output_file = output_file
         self._intermediate_dir = intermediate_dir
         self._rng = np.random.default_rng(seed=random_seed)
         np.random.seed(random_seed)
+        self._cores = cores
```
Comment (riemanli) [inline src/driver/experiment_driver.py:117]: Please modify the unit tests accordingly.

Comment (riemanli) [inline src/driver/experiment_driver.py:117]: Please modify the unit tests to accommodate the changes. Also, update the README if necessary.

**Review by riemanli — CHANGES_REQUESTED** [review]

**Comment by riemanli** [comment]
Hey Matthew, could you provide me an example command to run the `experiment_driver.py`, i.e. the full command of
```
python3 experiment_driver.py \
--data_design_dir=<data_design_dir> \
--experimental_design=<experimental_design> \
--output_file=<output_file> \
--intermediates_dir=<intermediates_dir> \
--seed=<random_seed> \
--cores=<number_of_cores>
```

File: src/driver/experiment_driver.py (hunk 1 above)
Comment (matthewclegg) [reply src/driver/experiment_driver.py:117]: I have added a unit test.  I will update the README in a separate PR.

**Review by jiayu-google — APPROVED** [review]
This feature is super helpful!  The codes are elegant.  Thank you Matt!

**Review by riemanli — APPROVED** [review]
LGTM.

Left an instruction of running the framework with multithreading for reference:
1. (Option) Decrease the number of datapoints and trials per design for a quick test:
    * `data_generators/single_publisher_design.py`: decrease the number of elements in `PUBLISHER_SIZES`
    * `driver/single_publisher_design.py`: change `NUM_TRIALS_PER_DATASET` to a smaller number.
2. Get synthetic data:
    * `python src/data_generators/synthetic_data_design_generator.py --output_dir=<output_dir_path> --data_design_config=single_publisher_design.py`
3. Run the evaluation:
    * `python src/driver/experiment_driver.py --data_design_dir=<synthetic_data_dir_path> --experimental_design=single_publisher_design.py --output_file=<output_file_path> --intermediates_dir=<intermediates_dir_path> --cores=0`

**Review by RonPinelliMRC — APPROVED** [review]
MRC approving based on cursory review after confirming non-independent code review on an exception basis. Experimental code will require independent review prior to publication of or reliance on results generated.

**Pull request merged**